packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java

```diff
@@ -64,16 +65,86 @@
 public class ImagePickerDelegate
     implements PluginRegistry.ActivityResultListener,
         PluginRegistry.RequestPermissionsResultListener {
-  private static final int REQUEST_CODE_CHOOSE_FROM_GALLERY = 2342;
-  private static final int REQUEST_CODE_TAKE_WITH_CAMERA = 2343;
+  static final int REQUEST_CODE_CHOOSE_FROM_GALLERY = 2342;
+  static final int REQUEST_CODE_TAKE_WITH_CAMERA = 2343;
 
-  private static final int REQUEST_EXTERNAL_STORAGE_PERMISSION = 2344;
-  private static final int REQUEST_CAMERA_PERMISSION = 2345;
+  static final int REQUEST_EXTERNAL_STORAGE_PERMISSION = 2344;
+  static final int REQUEST_CAMERA_PERMISSION = 2345;
 
   private final Activity activity;
   private final File externalFilesDirectory;
   private final ImageResizer imageResizer;
-  private final String providerName;
+
+  interface PermissionManager {
+    boolean isPermissionGranted(String permissionName);
+
+    void askForPermission(String permissionName, int requestCode);
+  }
+
+  interface IntentResolver {
+    boolean resolveActivity(Intent intent);
+  }
+
+  interface FileUriResolver {
+    Uri resolveFileProviderUriForFile(File imageFile);
+
+    void getFullImagePath(OnPathReadyListener listener);
+  }
+
+  interface OnPathReadyListener {
+    void onPathReady(String path);
+  }
+
+  @VisibleForTesting final String fileProviderName;
+
+  @VisibleForTesting
+  PermissionManager permissionManager =
+      new PermissionManager() {
+        @Override
+        public boolean isPermissionGranted(String permissionName) {
+          return ActivityCompat.checkSelfPermission(activity, permissionName)
+              == PackageManager.PERMISSION_GRANTED;
+        }
+
+        @Override
+        public void askForPermission(String permissionName, int requestCode) {
+          ActivityCompat.requestPermissions(activity, new String[] {permissionName}, requestCode);
+        }
+      };
+
+  @VisibleForTesting
+  IntentResolver intentResolver =
+      new IntentResolver() {
+        @Override
+        public boolean resolveActivity(Intent intent) {
+          return intent.resolveActivity(activity.getPackageManager()) != null;
+        }
+      };
+
+  @VisibleForTesting
+  FileUriResolver fileUriResolver =
```

Comment:
Suggestion: Create a second constructor for ImagePickerDelegate that (in addition to the already present arguments) takes an intentResolver, persmissionManager, etc. The existing Constructor would then instantiate these things as you do above and pass everything to the second new constructor. Clients would still use the existing Constructor, but in tests you can use the new constructor to pass in your mocks.

That way you avoid the slightly nasty global variables above.